core/src/banking_simulation.rs

```diff
@@ -0,0 +1,926 @@
+/// `SigVerifyStage`/gossip from a single thread to busy loop for precise T=N at ~1us granularity.
+///
+/// Warm-up starts at T=-WARMUP_DURATION (~ 13 secs). As soon as warm up is initiated, we invoke
+/// `BankingStage::new_num_threads()` as well to simulate the pre-leader slot's tx-buffering time.
+pub struct BankingSimulator {
+    banking_trace_events: BankingTraceEvents,
+    first_simulated_slot: Slot,
+}
+
+#[derive(Error, Debug)]
+pub enum SimulateError {
+    #[error("IO Error: {0}")]
+    IoError(#[from] io::Error),
+
+    #[error("Deserialization Error: {0}")]
+    SerializeError(#[from] bincode::Error),
+}
+
+// Defined to be enough to cover the holding phase prior to leader slots with some idling (+5 secs)
+const WARMUP_DURATION: Duration =
+    Duration::from_millis(HOLD_TRANSACTIONS_SLOT_OFFSET * DEFAULT_MS_PER_SLOT + 5000);
+
+/// BTreeMap is intentional because events could be unordered slightly due to tracing jitter.
+type PacketBatchesByTime = BTreeMap<SystemTime, (ChannelLabel, BankingPacketBatch)>;
+
+type FreezeTimeBySlot = BTreeMap<Slot, SystemTime>;
+
+type EventSenderThread = JoinHandle<(TracedSender, TracedSender, TracedSender)>;
+
+pub struct BankingTraceEvents {
+    packet_batches_by_time: PacketBatchesByTime,
+    freeze_time_by_slot: FreezeTimeBySlot,
+    hash_overrides: HashOverrides,
+}
+
+impl BankingTraceEvents {
+    fn read_event_file(
+        events: &mut Vec<TimedTracedEvent>,
+        event_file_path: &PathBuf,
+    ) -> Result<(), SimulateError> {
+        let mut reader = BufReader::new(File::open(event_file_path)?);
+
+        loop {
+            events.push(deserialize_from(&mut reader)?);
+
+            if reader.fill_buf()?.is_empty() {
+                // EOF is reached at a correct deserialization boundary.
+                // We're looking-ahead the buf, so NOT calling reader.consume(..) is correct.
+                break;
+            }
+        }
+
+        Ok(())
+    }
+
+    pub fn load(event_file_paths: &[PathBuf]) -> Result<Self, SimulateError> {
+        let mut events = vec![];
+        for event_file_path in event_file_paths {
+            let old_len = events.len();
+            let _ = Self::read_event_file(&mut events, event_file_path).inspect_err(|error| {
+                error!(
+                    "Reading {event_file_path:?} failed after {} events: {:?} due to file corruption or unclearn validator shutdown",
+                    events.len() - old_len,
+                    error
+                );
+            });
+            info!(
+                "Read {} events from {:?}",
+                events.len() - old_len,
+                event_file_path
+            );
+        }
+
+        let mut packet_batches_by_time = BTreeMap::new();
+        let mut freeze_time_by_slot = BTreeMap::new();
+        let mut hash_overrides = HashOverrides::default();
+        for TimedTracedEvent(event_time, event) in events {
```

Comment:
minor:
The collection of `events` seems unneccessary.

Why not just iterate over the traced events from reader directly, and then push directly to the appropriate data structure.

```rust
        let mut packet_batches_by_time = BTreeMap::new();
        let mut freeze_time_by_slot = BTreeMap::new();
        let mut hash_overrides = HashOverrides::default();
        for event_file_path in event_file_paths {
            let mut num_events = 0;
            let mut reader = BufReader::new(File::open(event_file_path)?);

            loop {
                let TimedTracedEvent(event_time, event): TimedTracedEvent =
                    deserialize_from(&mut reader)?;
                num_events += 1;

                match event {
                    TracedEvent::PacketBatch(label, batch) => {
                        // Deserialized PacketBatches will mostly be ordered by event_time, but this
                        // isn't guaranteed when traced, because time are measured by multiple _sender_
                        // threads withtout synchnotization among them to avoid overhead.
                        //
                        // Also, there's a possiblity of system clock change. In this case,
                        // the simulation is meaningless, though...
                        //
                        // Somewhat naively assume that event_times (nanosecond resolution) won't
                        // collide.
                        let is_new = packet_batches_by_time
                            .insert(event_time, (label, batch))
                            .is_none();
                        assert!(is_new);
                    }
                    TracedEvent::BlockAndBankHash(slot, blockhash, bank_hash) => {
                        let is_new = freeze_time_by_slot.insert(slot, event_time).is_none();
                        hash_overrides.add_override(slot, blockhash, bank_hash);
                        assert!(is_new);
                    }
                }

                if reader.fill_buf()?.is_empty() {
                    // EOF is reached at a correct deserialization boundary.
                    // We're looking-ahead the buf, so NOT calling reader.consume(..) is correct.
                    break;
                }
            }

            info!("Read {num_events} events from {event_file_path:?}");
        }
```


could probably create an iterator out of each file, but i didn't figure it out here.

Reply:
oh, thanks for good catch. yeah, `events` are some relic of past hacky impl...: af866619cff63c76774db3d0d43695fffc6c04f9 (I went with plain old closures, instead of iterators to avoid coupling)

> could probably create an iterator out of each file, but i didn't figure it out here.

fyi, i found https://crates.io/crates/concat-reader, but opted not to use it because rather clumsy error handling.